File: sdk/typescript/src/batch.ts
/**
 * ATLAST ECP Batch — collect records and upload to backend
 */

import { buildMerkleRoot, signData } from './crypto';
import { collectBatch } from './storage';
import { loadOrCreateIdentity } from './identity';
import type { BatchUploadRequest, BatchUploadResponse, ATLASTConfig, ECPRecord } from './types';

const DEFAULT_API_URL = process.env.ATLAST_API_URL || '';  // User must configure

export async function uploadBatch(config: ATLASTConfig): Promise<BatchUploadResponse> {
  const identity = loadOrCreateIdentity(config.agentId);
  const { records, hashes } = collectBatch(config.agentId, config.batchSize || 100);

  if (records.length === 0) {
    return { batch_id: '', status: 'empty', message: 'No records to upload', attestation_uid: undefined };
  }

  const merkleRoot = buildMerkleRoot(hashes);
  const sig = signData(identity.private_key, merkleRoot);

  // Compute flag counts
  const flagCounts: Record<string, number> = {};
  let totalLatency = 0;
  let latencyCount = 0;
  for (const r of records) {
    for (const f of r.flags) {
      flagCounts[f] = (flagCounts[f] || 0) + 1;
    }
    if (r.latency_ms) {
      totalLatency += r.latency_ms;
      latencyCount++;
    }
  }

  const avgLatency = latencyCount > 0 ? Math.round(totalLatency / latencyCount) : 0;

  const payload: BatchUploadRequest = {
    agent_did: identity.did,
    merkle_root: merkleRoot,
    record_count: records.length,
    avg_latency_ms: avgLatency,
    batch_ts: Date.now(),
    sig,
    ecp_version: '0.5',
    flag_counts: flagCounts,
    record_hashes: records.map((r) => ({
      id: r.id,
      hash: r.chain.hash,
      flags: r.flags,
    })),
  };

  const apiUrl = config.apiUrl || process.env.ATLAST_API_URL || DEFAULT_API_URL;
  const apiKey = config.apiKey || process.env.ATLAST_API_KEY;
  const headers: Record<string, string> = { 'Content-Type': 'application/json' };
  if (apiKey) {
    headers['X-Agent-Key'] = apiKey;
  }

  const resp = await fetch(`${apiUrl}/v1/batches`, {
    method: 'POST',
    headers,
    body: JSON.stringify(payload),
  });

  if (!resp.ok) {
    const errText = await resp.text();
    throw new Error(`Batch upload failed (${resp.status}): ${errText}`);
  }

  return (await resp.json()) as BatchUploadResponse;
}

File: sdk/go/batch.go
package ecp

import (
	"bytes"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"
)

// Batch represents a batch of ECP records ready for upload.
type Batch struct {
	ID           string   `json:"id"`
	MerkleRoot   string   `json:"merkle_root"`
	RecordCount  int      `json:"record_count"`
	RecordHashes []string `json:"record_hashes"`
	CreatedAt    int64    `json:"created_at"`
}

// BuildBatch creates a Batch from a slice of records.
// The MerkleRoot is computed from each record's chain hash (or computed hash if chain is absent).
func BuildBatch(records []Record) Batch {
	hashes := make([]string, 0, len(records))
	for _, r := range records {
		h := ""
		if r.Chain != nil && r.Chain.Hash != "" {
			h = r.Chain.Hash
		} else {
			h = ComputeChainHash(r)
		}
		hashes = append(hashes, h)
	}

	root := BuildMerkleRoot(hashes)

	b := make([]byte, 8)
	_, _ = rand.Read(b)

	return Batch{
		ID:           "batch_" + hex.EncodeToString(b),
		MerkleRoot:   root,
		RecordCount:  len(records),
		RecordHashes: hashes,
		CreatedAt:    time.Now().UnixMilli(),
	}
}

// UploadBatch POSTs a batch to the ATLAST API.
// apiURL should be the full versioned base URL, e.g. "https://api.weba0.com/v1".
// apiKey is sent as X-API-Key header (omitted if empty).
// agentDID and sig are required by the server.
func UploadBatch(apiURL, apiKey, agentDID, sig string, batch Batch) error {
	payload, err := json.Marshal(map[string]interface{}{
		"merkle_root":   batch.MerkleRoot,
		"agent_did":     agentDID,
		"record_count":  batch.RecordCount,
		"record_hashes": batch.RecordHashes,
		"batch_ts":      batch.CreatedAt,
		"sig":           sig,
		"ecp_version":   "0.1",
	})
	if err != nil {
		return err
	}

	req, err := http.NewRequest(http.MethodPost, apiURL+"/batches", bytes.NewReader(payload))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")
	if apiKey != "" {
		req.Header.Set("X-API-Key", apiKey)
	}

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	io.Copy(io.Discard, resp.Body) //nolint

	if resp.StatusCode >= 400 {
		return fmt.Errorf("upload failed: HTTP %d", resp.StatusCode)
	}
	return nil
}
